### DISCUSSION TITLE
Lyra Intergration

### DISCUSSION THREAD

**Discussion by ZacharyDK** [Sixze/ALS-Refactored#328] [General]
Been working on how to best get ALS into Lyra. I have a successful partial integration I'd like to share. It is possible to have ALS use Lyra's weapon framework.

1. Use this version of ALS. Keep it as a plugin.
2. ALS char now needs to inherit from AModularCharacter. ModularGameplay and ModularGameplay actors are now public dependencies. This is necessary because Lyra adds an EquipmentManagerComponent to LyraCharacter at runtime via experiences/modular gameplay
3. I tried to re parent ALS character to Lyra character. Broke ALS animation. Still not sure why,suspect it is related to equipment.
4. Couldn't get ILyraTeamAgentInterface to work. Some compiler linker error that even I don't understand. 
5. Lyra directly sets linked anim proxys with the equipment. ALS changes overlay states by handling when a gameplay tag has changed. Override onequip and unequip functions in the LyraWeaponInstance (can be done in BP) and use an interface to send tag info to the pawn. Don't directly changed what anims layers are linked. Though this ALS's architecture is very similar to Lyra, more work/testing  needs to be done to handle directly managing this. 
6. You need to know how experiences are loaded. Things won't work without a proper gamemode. See. https://x157.github.io/UE5/LyraStarterGame/Experience/#LyraGameMode  
 
You have to do your reading, because there is a lack of helpful debug messages...like if you were to try to modular gameplay stuff on an actor that doesn't inherit from modular gameplay actor...

7. You can just use the ALS character, but use Lyra controller, Gamemode etc. Not sure why inheriting from ALSExampleCharacter is borked. 

<img width="1001" alt="Screenshot 2023-06-30 at 10 13 15 AM" src="https://github.com/Sixze/ALS-Refactored/assets/40034922/e3b6c4a9-ff3f-41aa-9950-0d93b2699399">
<img width="736" alt="Screenshot 2023-06-30 at 10 10 58 AM" src="https://github.com/Sixze/ALS-Refactored/assets/40034922/0a49b202-c1ec-48cf-9f17-0423ea0446af">
<img width="685" alt="Screenshot 2023-06-30 at 10 09 25 AM" src="https://github.com/Sixze/ALS-Refactored/assets/40034922/3a22df01-a5c5-4ae0-8ec9-34605203a2b4">
<img width="1792" alt="Screenshot 2023-06-30 at 10 08 34 AM" src="https://github.com/Sixze/ALS-Refactored/assets/40034922/d5fea95b-7edc-4958-9c02-cb13f2252b47">

**Comment by N0mN0mT0m** [comment]
Greetings @ZacharyDK ,
Did you make any progress with the botched animations when parenting from ALyraCharacter,
I am trying to do the same, where I have parented my AALSCharacter from ALyraCharacter and can confirm that the inputs are going through from my LyraPlayerController/LyraHeroComponent , where things like jump and crouch are working well assuming that it sits in the same ACharacter framework, there are issues triggering any other animations from ALS, or having the move animations trigger correctly.
Any insight would be really helpful.

**Reply by ZacharyDK** [reply]
Nope. I couldn't figure out why it was borked. I decided it would be best to forgo any ALS integration and just use Lyra and tack on the systems from ALS i liked. 


The reason I wanted to use ALS in the first place was because I liked the sprint and montage and mantle system. Lyra has its own montage stuff, and it was easier to hack together a sprint system from ALS's work. Still haven't totally ported a mantle system over.

I have a half decent sprint system by 

1. Override ULyraCharacterMovementComponent::GetMaxSpeed() . Check gameplay tags on ASC. We set maxspeed to sprintspeed sprint if CanSprint() is true and have the TAG_Gameplay_MovementSprint. (Or whatever you called your sprint tag)
2. In ALyraCharacter tick function, Speed = GetVelocity().Length. if/el if statements comparing speed to StandingMaxSpeed, WalkingMaxSpeed, runningMaxspeed, and sprint max speed. Also get InputDirectionYawAngle, InputRelativeTransform
om gameplay tags for standing,walking,running,etc SetLooseGameplayTagCount accordingly .    
3. CanSprint in ALyraChatcter


`if(ViewRelativeAngleThreshold*-1.0f <= InputRelativeTransform.Rotator().Yaw && InputRelativeTransform.Rotator().Yaw <= ViewRelativeAngleThreshold && !bIsCrouched)
	{
		return true;
	}
return false'


Tick 

`void ALyraCharacter::Tick(const float DeltaTime)
{
	//Refresh Input
	FVector_NetQuantizeNormal NewInputDirection = (GetCharacterMovement()->GetCurrentAcceleration() / GetCharacterMovement()->GetMaxAcceleration());
	if (GetLocalRole() >= ROLE_AutonomousProxy)
	{
		NewInputDirection.GetSafeNormal();
		//COMPARE_ASSIGN_AND_MARK_PROPERTY_DIRTY(ThisClass, InputDirection, InputDirection, this);
		InputDirection = NewInputDirection;
	}
	bHasInput = InputDirection.SizeSquared() > UE_KINDA_SMALL_NUMBER;
	//Input direction
	if (bHasInput)
	{
		InputDirectionYawAngle = UE_REAL_TO_FLOAT(FMath::RadiansToDegrees(FMath::Atan2(InputDirection.Y, InputDirection.X)));
	}

	FTransform A = FTransform(FRotator(0.0f,InputDirectionYawAngle,0.0f),FVector(0.0f,0.0f,0.0f),FVector(1.0f,1.0f,1.0f));
	InputRelativeTransform = UKismetMathLibrary::MakeRelativeTransform(A,GetActorTransform());

	if (ULyraAbilitySystemComponent* LyraASC = GetLyraAbilitySystemComponent())
	{
		float Speed = GetVelocity().Length();  //GetCharacterMovement()->GetVelo

		if (Speed <= StandingMaxSpeed)
		{
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Standing, 1);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Walking, 0);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Running, 0);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Sprinting, 0);


		} else if (Speed > StandingMaxSpeed && Speed <= WalkingMaxSpeed)
		{
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Standing, 0);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Walking, 1);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Running, 0);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Sprinting, 0);

		} else if (Speed  > WalkingMaxSpeed && Speed <= RunningMaxSpeed )
		{
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Standing, 0);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Walking, 0);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Running, 1);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Sprinting, 0);

		} else if (Speed > RunningMaxSpeed)
		{
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Standing, 0);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Walking, 0);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Running, 0);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Sprinting, 1);
		}

		if (GetCharacterMovement()->IsFalling())
		{
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Grounded, 0);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Air, 1);
		}
		else
		{
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Grounded, 1);
			LyraASC->SetLooseGameplayTagCount(LyraGameplayTags::Status_Air, 0);
		}


	}


	Super::Tick(DeltaTime);
}`
